nav2_costmap_2d/src/clear_costmap_service.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <memory>
+
+#include "nav2_costmap_2d/clear_costmap_service.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+namespace nav2_costmap_2d
+{
+
+using std::vector;
+using std::string;
+using std::shared_ptr;
+using std::any_of;
+using ClearExceptRegion = nav2_msgs::srv::ClearCostmapExceptRegion;
+
+ClearCostmapService::ClearCostmapService(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
+: node_(node), costmap_(costmap)
+{
+  node_->get_parameter_or_set("clearable_layers", clearable_layers_, {"obstacle_layer"});
+
+  server_ = node_->create_service<ClearExceptRegion>("clear_except_" + costmap->getName(),
+      std::bind(&ClearCostmapService::clearExceptRegionCallback, this,
+      std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+}
+
+void ClearCostmapService::clearExceptRegionCallback(
+  const shared_ptr<rmw_request_id_t>/*request_header*/,
+  const shared_ptr<ClearExceptRegion::Request> request,
+  const shared_ptr<ClearExceptRegion::Response>/*response*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to clear " + costmap_->getName());
+
+  if (costmap_ == nullptr) {
```

Comment:
If you change the constructor to take a `Costmap2DROS & costmap` instead, then you don't need this check.

Reply:
Good point, using a reference makes more sense in this case.